dev/com.ibm.ws.st.liberty.buildplugin.integration/src/com/ibm/ws/st/liberty/buildplugin/integration/internal/AbstractLibertyRuntimeProvider.java

```diff
@@ -0,0 +1,304 @@
+            }
+            return null;
+        }
+        return config;
+    }
+
+    /*
+     * @param project - The Project containing the Liberty Runtime
+     *
+     * @return UserDirectory
+     */
+    private UserDirectory getUserDirectory(IProject project, LibertyBuildPluginConfiguration config) {
+        // Find the install dir of the runtime
+        String installDir = config.getConfigValue(ConfigurationType.installDirectory);
+
+        WebSphereRuntime wsRuntime = getWSRuntime(installDir);
+
+        String userDirVal = config.getConfigValue(ConfigurationType.userDirectory);
+        if (userDirVal != null && wsRuntime != null) {
+            IPath userPath = new Path(userDirVal);
+            return new UserDirectory(wsRuntime, userPath, project);
+        }
+        return null;
+    }
+
+    private WebSphereRuntime getWSRuntime(IProject project, LibertyBuildPluginConfiguration config) {
+        // Find the install dir of the runtime
+        String installDir = config.getConfigValue(ConfigurationType.installDirectory);
+        if (installDir == null) {
+            return null;
+        }
+        return getWSRuntime(installDir);
+    }
+
+    private WebSphereRuntime getWSRuntime(String installDir) {
+
+        // Check if the runtime already exists
+        IRuntime[] runtimes = ServerCore.getRuntimes();
+        IPath installDirectory = new Path(installDir);
+        IRuntime runtime = null;
+        for (IRuntime rt : runtimes) {
+            if (WebSphereUtil.isWebSphereRuntime(rt) && installDirectory.equals(rt.getLocation())) {
+                runtime = rt;
+            }
+        }
+        WebSphereRuntime wsRuntime = null;
+        if (runtime != null) {
+            wsRuntime = (WebSphereRuntime) runtime.loadAdapter(WebSphereRuntime.class, null);
+        } else {
+            // If there is no runtime created in the workspace, then provide our own ghost runtime 'instance'
+            IRuntimeWorkingCopy runtimeFromDir = WebSphereRuntimeLocator.getRuntimeFromDir(installDirectory, new NullProgressMonitor());
+            if (runtimeFromDir != null) {
+                wsRuntime = (WebSphereRuntime) runtimeFromDir.loadAdapter(WebSphereRuntime.class, null);
+            }
+        }
+        return wsRuntime;
+    }
+
+    /**
+     * Returns the path to the 'actual' server.xml in the target or build folder, where the embedded runtime is
+     *
+     * @param serverOutputDirectory
+     * @return
+     */
+    private String getTargetServerConfig(String serverOutputDirectory) {
+        return serverOutputDirectory + "/server.xml";
+    }
+
+    private IResource convertConfigValueToIResource(String path, IProject project) {
+        try {
+            URI pathURI = URI.create(path.replaceAll(" ", "%20")); // Encode spaces in path
```

Comment:
I think it would be safer to create a File and then call File.toURI().  I got an exception here right away: java.lang.IllegalArgumentException: Illegal character in opaque part at index 2: C:\home\work\gradle\Web1\build\wlp\usr\servers\defaultServer/server.xml